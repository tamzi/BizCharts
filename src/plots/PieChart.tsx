import 'react';
import { Pie, PieOptions as options } from '@antv/g2plot/lib/plots/pie';
import createPlot, { BasePlotOptions } from '../createPlot';
import { polyfillOptions } from './core/polyfill';
import { LengendAPIOptions, TooltipAPIOptions, LabelAPIOptions } from './core/interface';

interface PieOptions extends options, BasePlotOptions {
    tooltip?: TooltipAPIOptions,
    label?: LabelAPIOptions,
    legend?: LengendAPIOptions,
};

export { PieOptions };
export default createPlot<PieOptions>(Pie, 'PieChart', polyfillOptions);
