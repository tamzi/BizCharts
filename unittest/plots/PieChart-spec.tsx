import React, { useState } from 'react';
import PieChart from '../../src/plots/PieChart';
import { render, cleanup } from '@testing-library/react';

const MOCK_DATA = [
  {
    type: '分类一',
    value: 2745678987654323456789098765432345678,
  },
  {
    type: '分类二',
    value: 252345678909876543456789987654345678,
  },
  {
    type: '分类三',
    value: 345678998765434567876545678876345678,
  },
  {
    type: '分类四',
    value: 234567809876544567890123456788765678,
  },
  {
    type: '分类五',
    value: 234567854345678987655678987655678987,
  },
  {
    type: '其它',
    value: 123456789876789098767898765678909876,
  },
];

const padding = [90, 90, 90, 90];

const basicCfg = {
  padding: padding,
  data: MOCK_DATA,
  angleField: "value",
  colorField: "type",
  forceFit: true,
  width: 500,
  height: 500,
}
const Demo = (props) => {
  const [state, setState] = useState(1);

  return (<div>
    <a onClick={() => setState(state + 1)}>
      click:{state}
    </a>
    <PieChart
      {...basicCfg}
      title="click 后动画不更新"
      {...props}
    />
  </div >)
}

describe('Plots-PieChart', () => {
  test('click 后动画不更新', () => {
    render(<Demo />);
    cleanup();
  });

  test('基础饼图', () => {
    let chart = null;
    render(<PieChart
      {...basicCfg}
      title="基础饼图"
      pixelRatio={3}
      renderer="svg"
      meta={{
        type: {
          formatter: val => `${val}%`
        }
      }}
      radius={0.8}
      // innerRadius={0.6}
      color={['red', 'green', 'yellow','black','orange']}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });


  test('legend-饼图:text转itemName', () => {
    const fn = jest.fn(val => `${val}%`);
    const fn2 = jest.fn(val => `${val}%`);
    let chart = null;
    render(<PieChart
      title="legend-饼图"
      {...basicCfg}
      legend={{
        visible: true,
        position: "bottom",
        formatter: fn, //  text 的优先度大于formatter
        offsetX: 40,
        offsetY: 2,
        title: {
          visible: true,
          text: 'legend tile',
          style: {
            fill: "red",
          },
        },
        marker: {
          symbol: "circle",
        },
        text: {
          style: {
            fill: "pink",
            stroke: "pink",
          },
          formatter: fn2,
        },
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(fn).not.toHaveBeenCalled();
    expect(fn2).toHaveBeenCalled();
    cleanup();
  });


  test('tooltip饼图', () => {
    let chart = null;
    render(<PieChart
      title="tooltip-饼图"
      {...basicCfg}
      onGetG2Instance={
        (c) => chart = c
      }
      tooltip={{
        visible: true,
        offset: 10,
        // shared: true,
        showCrosshairs: true,
        crosshairs: { type: 'y' },
        fields: ['type', 'value'],
        formatter: ({ type, value }) => ({ value, name: `${type}-tyz` })
      }}
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('label饼图', () => {
    let chart = null;
    const opt = {
      data: MOCK_DATA,
      width: 500,
      height: 500,
      angleField: 'value',
      colorField: 'type',
      autoFit: false,
      radius: 0.4,
      label: {
        type: 'outer',
        autoRotate: true,
        formatter: ({ value }) => `val-${value}`,
        offsetX: 6,
        offsetY: 6,
        style: {
          fill: 'red',
        },
        autoHide: false,
      }
    }
    render(<PieChart
      title="label饼图"
      {...opt}
      onGetG2Instance={
        (c) => {console.log(c); chart = c}
      }
    />);
    expect(chart.options).toMatchSnapshot();
    // cleanup();
  });
});
