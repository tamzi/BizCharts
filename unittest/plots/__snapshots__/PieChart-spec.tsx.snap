// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plots-PieChart label饼图 1`] = `
Object {
  "angleField": "value",
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": false,
  "colorField": "type",
  "data": Array [
    Object {
      "type": "分类一",
      "value": 2.7456789876543237e+36,
    },
    Object {
      "type": "分类二",
      "value": 2.5234567890987654e+35,
    },
    Object {
      "type": "分类三",
      "value": 3.456789987654346e+35,
    },
    Object {
      "type": "分类四",
      "value": 2.345678098765446e+35,
    },
    Object {
      "type": "分类五",
      "value": 2.3456785434567898e+35,
    },
    Object {
      "type": "其它",
      "value": 1.234567898767891e+35,
    },
  ],
  "height": 500,
  "label": Object {
    "autoHide": false,
    "autoRotate": true,
    "formatter": [Function],
    "layout": Object {
      "cfg": Object {
        "action": "ellipsis",
      },
      "type": "limit-in-plot",
    },
    "offsetX": 6,
    "offsetY": 6,
    "style": Object {
      "fill": "red",
    },
    "type": "outer",
  },
  "legend": Object {
    "position": "right",
  },
  "pieStyle": Object {
    "lineWidth": 1,
    "stroke": "white",
  },
  "radius": 0.4,
  "renderer": "canvas",
  "statistic": Object {
    "content": Object {
      "style": Object {
        "color": "rgba(44,53,66,0.85)",
        "fontSize": "32px",
        "fontWeight": "bold",
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
    "title": Object {
      "style": Object {
        "color": "#4B535E",
        "fontSize": "20px",
        "fontWeight": 300,
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
  },
  "theme": Object {
    "components": Object {
      "annotation": Object {
        "text": Object {
          "animate": false,
        },
      },
    },
  },
  "tooltip": Object {
    "shared": false,
    "showMarkers": false,
    "showTitle": false,
  },
  "width": 500,
  "xAxis": Object {
    "label": Object {
      "autoHide": Object {
        "cfg": Object {
          "minGap": 6,
        },
        "type": "equidistance",
      },
      "autoRotate": false,
    },
    "nice": true,
  },
  "yAxis": Object {
    "label": Object {
      "autoHide": true,
      "autoRotate": false,
    },
    "nice": true,
  },
}
`;

exports[`Plots-PieChart tooltip饼图 1`] = `
Object {
  "angleField": "value",
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "colorField": "type",
  "data": Array [
    Object {
      "type": "分类一",
      "value": 2.7456789876543237e+36,
    },
    Object {
      "type": "分类二",
      "value": 2.5234567890987654e+35,
    },
    Object {
      "type": "分类三",
      "value": 3.456789987654346e+35,
    },
    Object {
      "type": "分类四",
      "value": 2.345678098765446e+35,
    },
    Object {
      "type": "分类五",
      "value": 2.3456785434567898e+35,
    },
    Object {
      "type": "其它",
      "value": 1.234567898767891e+35,
    },
  ],
  "height": 500,
  "label": Object {
    "layout": Object {
      "cfg": Object {
        "action": "ellipsis",
      },
      "type": "limit-in-plot",
    },
  },
  "legend": Object {
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "pieStyle": Object {
    "lineWidth": 1,
    "stroke": "white",
  },
  "renderer": "canvas",
  "statistic": Object {
    "content": Object {
      "style": Object {
        "color": "rgba(44,53,66,0.85)",
        "fontSize": "32px",
        "fontWeight": "bold",
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
    "title": Object {
      "style": Object {
        "color": "#4B535E",
        "fontSize": "20px",
        "fontWeight": 300,
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
  },
  "theme": Object {
    "components": Object {
      "annotation": Object {
        "text": Object {
          "animate": false,
        },
      },
    },
  },
  "tooltip": Object {
    "crosshairs": Object {
      "type": "y",
    },
    "fields": Array [
      "type",
      "value",
    ],
    "formatter": [Function],
    "offset": 10,
    "shared": false,
    "showCrosshairs": true,
    "showMarkers": false,
    "showTitle": false,
    "visible": true,
  },
  "width": 500,
  "xAxis": Object {
    "label": Object {
      "autoHide": Object {
        "cfg": Object {
          "minGap": 6,
        },
        "type": "equidistance",
      },
      "autoRotate": false,
    },
    "nice": true,
  },
  "yAxis": Object {
    "label": Object {
      "autoHide": true,
      "autoRotate": false,
    },
    "nice": true,
  },
}
`;

exports[`Plots-PieChart 基础饼图 1`] = `
Object {
  "angleField": "value",
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "color": Array [
    "red",
    "green",
    "yellow",
    "black",
    "orange",
  ],
  "colorField": "type",
  "data": Array [
    Object {
      "type": "分类一",
      "value": 2.7456789876543237e+36,
    },
    Object {
      "type": "分类二",
      "value": 2.5234567890987654e+35,
    },
    Object {
      "type": "分类三",
      "value": 3.456789987654346e+35,
    },
    Object {
      "type": "分类四",
      "value": 2.345678098765446e+35,
    },
    Object {
      "type": "分类五",
      "value": 2.3456785434567898e+35,
    },
    Object {
      "type": "其它",
      "value": 1.234567898767891e+35,
    },
  ],
  "height": 500,
  "label": Object {
    "layout": Object {
      "cfg": Object {
        "action": "ellipsis",
      },
      "type": "limit-in-plot",
    },
  },
  "legend": Object {
    "position": "right",
  },
  "meta": Object {
    "type": Object {
      "formatter": [Function],
    },
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "pieStyle": Object {
    "lineWidth": 1,
    "stroke": "white",
  },
  "pixelRatio": 3,
  "radius": 0.8,
  "renderer": "svg",
  "statistic": Object {
    "content": Object {
      "style": Object {
        "color": "rgba(44,53,66,0.85)",
        "fontSize": "32px",
        "fontWeight": "bold",
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
    "title": Object {
      "style": Object {
        "color": "#4B535E",
        "fontSize": "20px",
        "fontWeight": 300,
        "lineHeight": 1,
        "textAlign": "center",
      },
    },
  },
  "theme": Object {
    "components": Object {
      "annotation": Object {
        "text": Object {
          "animate": false,
        },
      },
    },
  },
  "tooltip": Object {
    "shared": false,
    "showMarkers": false,
    "showTitle": false,
  },
  "width": 500,
  "xAxis": Object {
    "label": Object {
      "autoHide": Object {
        "cfg": Object {
          "minGap": 6,
        },
        "type": "equidistance",
      },
      "autoRotate": false,
    },
    "nice": true,
  },
  "yAxis": Object {
    "label": Object {
      "autoHide": true,
      "autoRotate": false,
    },
    "nice": true,
  },
}
`;
